### PR TITLE
[Fix/CON-631]: update tests for navigation desktop version

### PR DESCRIPTION
**Pull request by hassanraza0054**
Ticket: https://aesoponline.atlassian.net/browse/CON-631

### CODE REVIEW

File: src/components/GlobalNavigation/components/DesktopView/components/CollectionItem/CollectionItem.spec.tsx [hunk 1]
```diff
@@ -1,38 +1,150 @@
...
+      <BaseComponentWithProviders
+        {...{ ...CollectionItemLinkFixture, isExternal: true }}
+      />,
+    );
+
+    const results = await axe(container);
+    const link = screen.getByRole('link');
+
+    expect(results).toHaveNoViolations();
+    expect(link).not.toHaveAttribute('tabIndex');
```
Comment (MkMan) [inline src/components/GlobalNavigation/components/DesktopView/components/CollectionItem/CollectionItem.spec.tsx:115]: I am struggling to make the connection between _being tabbable_ and not having a `tabIndex`. Can you please elaborate?

File: src/components/GlobalNavigation/components/DesktopView/components/CollectionItem/CollectionItem.spec.tsx [hunk 2]
```diff
@@ -1,38 +1,150 @@
...
+    const results = await axe(container);
+    const link = screen.queryByRole('link');
+
+    link.click();
+    userEvent.hover(link);
 
-  //   expect(results).toHaveNoViolations();
-  // });
+    expect(results).toHaveNoViolations();
+    expect(link).toHaveAttribute('data-test-id', CollectionItemLinkFixture.id);
```
Comment (MkMan) [inline src/components/GlobalNavigation/components/DesktopView/components/CollectionItem/CollectionItem.spec.tsx:147]: This assertion doesn't seem related to test description. You could change how you get a handle to the link from `byRole` to `byTestId` but then that element could be anything and not a link.

File: src/components/GlobalNavigation/components/DesktopView/components/CollectionLayout/CollectionLayout.spec.tsx [hunk 3]
```diff
@@ -1,51 +1,101 @@
...
   });
+
+  it('should render a read collection', () => {
+    render(
+      <BaseComponentWithProviders
+        collectionLayout={ReadCollectionLayoutFixture}
+      />,
+    );
+
+    const small = screen.queryByText(
```
Comment (MkMan) [inline src/components/GlobalNavigation/components/DesktopView/components/CollectionLayout/CollectionLayout.spec.tsx:75]: `small` is a bit generic as a variable name. Can you please make it more contextual?

File: src/components/GlobalNavigation/components/DesktopView/components/CollectionLayout/CollectionLayout.spec.tsx [hunk 4]
```diff
@@ -1,51 +1,101 @@
...
+      <BaseComponentWithProviders
+        collectionLayout={OtherCollectionLayoutFixture}
+      />,
+    );
+
+    const promoCard = screen.getByTitle(
+      OtherCollectionLayoutFixture.promotion.title,
+    );
+
+    expect(promoCard).not.toHaveAttribute('tabIndex');
```
Comment (MkMan) [inline src/components/GlobalNavigation/components/DesktopView/components/CollectionLayout/CollectionLayout.spec.tsx:99]: This assertion doesn't match the test description. Would `toBeInTheDocument` be a more suitable choice here?

**Review by MkMan — COMMENTED** [review]
Good stuff 👌 

I noticed you use both `getBy...` and `queryBy...`. This [table](https://testing-library.com/docs/react-testing-library/cheatsheet/#queries) goes through the differences between the 3 (get, query, find).

My approach is:
- use _get_ as a first preference
- use _query_ to "find" things not in the DOM and then assert the element is `not.toBeInTheDocument` as get would throw an error
- use _find_ to find elements that may require waiting for asynchronous code to finish before being rendered

Keen to hear your thoughts and view on which to use when 😃

**Comment by cyrusxzw** [comment]
> queryBy

I think when checking if `toBeInTheDocument()`, better to use `queryBy`. Because `getBy `throws an error if the element is not found, but `queryBy` doesn't and continue the test.

File: src/components/GlobalNavigation/components/DesktopView/components/CloseButton/CloseButton.spec.tsx [hunk 5]
```diff
@@ -1,37 +1,63 @@
...
+    (useGlobalNavigationStateContext as jest.Mock).mockReturnValue({
+      ...testGlobalNavigationStateContext,
+    });
+  });
+
+  it('should render base component correctly', async () => {
+    const { container } = render(<BaseComponentWithProviders />);
+
+    expect(container).toMatchSnapshot();
+  });
```
[the remark this answers is not in the document]
Comment (hassanraza0054) [reply src/components/GlobalNavigation/components/DesktopView/components/CloseButton/CloseButton.spec.tsx:50]: > Good stuff 👌
> 
> I noticed you use both `getBy...` and `queryBy...`. This [table](https://testing-library.com/docs/react-testing-library/cheatsheet/#queries) goes through the differences between the 3 (get, query, find).
> 
> My approach is:
> 
> * use _get_ as a first preference
> * use _query_ to "find" things not in the DOM and then assert the element is `not.toBeInTheDocument` as get would throw an error
> * use _find_ to find elements that may require waiting for asynchronous code to finish before being rendered
> 
> Keen to hear your thoughts and view on which to use when 😃

Nah that's perfect. I was confused whether to use `getBy` or `queryBy` and chose `queryBy` initially as it doesn't return an error in the test. Curious to know how you developed your approach with regards to findBy, getBy and queryBy. 

Till then, thanks for the review. Gonna make the suggested changes and answer comments where I need a bit more input!👌

[the remark this answers is not in the document]
Comment (MkMan) [reply src/components/GlobalNavigation/components/DesktopView/components/CloseButton/CloseButton.spec.tsx:50]: > Curious to know how you developed your approach with regards to findBy, getBy and queryBy

It started with me focusing on just using 1 of them. I liked `get` because it throws
* if the element isn't found
* if multiple are found

I felt that made it safer to use mainly because of the last point. It meant that if the DOM had multiple elements that matched the selector you'd be alerted (through a failing test) instead of being allowed to use the wrong one, thus saving you debugging time.

However, using `get` alone isn't sufficient. Because it will throw if it can't find an element that meant I couldn't use it to assert elements were **not** in the DOM. I sometimes use this assertion to test elements that are rendered conditionally. So for that I use `query` as it won't throw.

Lastly, neither `get` nor `query` will wait for an element to be found so when I have a situation that requires waiting I use `for`.

File: jest.config.js [hunk 6]
```diff
@@ -11,7 +11,7 @@ module.exports = {
     'fixture.(jsx?|tsx?)',
     '.storybook',
     'index.(js|ts)',
-    'components/GlobalNavigation' /* TODO: add full test coverage for Global Navigation */,
+    'components/GlobalNavigation/components/TabletView' /* TODO: add full test coverage for Global Navigation */,
```
Comment (cyrusxzw) [inline jest.config.js:14]: Ignore MobileView too?

File: src/components/GlobalNavigation/components/DesktopView/components/CloseButton/CloseButton.spec.tsx [hunk 7]
```diff
@@ -1,37 +1,63 @@
...
+} from '~/components/GlobalNavigation/GlobalNavigation.context';
+import { GlobalNavigationFixture } from '~/components/GlobalNavigation/GlobalNavigation.fixture';
+
+import { CloseButton } from './CloseButton';
+import { CloseButtonFixture } from './CloseButton.fixture';
+
+jest.mock('~/components/GlobalNavigation/GlobalNavigation.context');
+
+const BaseComponentWithProviders = () => (
+  <GlobalNavigationStateContextProvider activeCollectionId="home" isOpen={true}>
```
Comment (cyrusxzw) [inline src/components/GlobalNavigation/components/DesktopView/components/CloseButton/CloseButton.spec.tsx:17]: I feel like `activeCollectionId="top"` not 'home'. Correct me if i'm wrong.

File: src/components/GlobalNavigation/components/DesktopView/components/CloseButton/CloseButton.spec.tsx [hunk 8]
```diff
@@ -1,37 +1,63 @@
...
+    });
+  });
+
+  it('should render base component correctly', async () => {
+    const { container } = render(<BaseComponentWithProviders />);
+
+    expect(container).toMatchSnapshot();
+  });
+
+  it('should be tabbable', async () => {
```
Comment (cyrusxzw) [inline src/components/GlobalNavigation/components/DesktopView/components/CloseButton/CloseButton.spec.tsx:52]: This is for a11y test, right? Why not use `axe`?

**Comment by cyrusxzw** [comment]
Nice work 👍!

File: jest.config.js (hunk 6 above)
Comment (hassanraza0054) [reply jest.config.js:14]: Nope we already have mobile tests! The only one left is tablet view now. So keeping that ignored for now until we add coverage there.

File: src/components/GlobalNavigation/components/DesktopView/components/CloseButton/CloseButton.spec.tsx (hunk 8 above)
Comment (hassanraza0054) [reply src/components/GlobalNavigation/components/DesktopView/components/CloseButton/CloseButton.spec.tsx:52]: Haven't used a specific test for axe. I know we have `x.toHaveNoViolations()` but when I try that, I get these various errors.

image

I'm not sure whats the best way to go about fixing these, and hope to discuss with the team in some future session as to how we can go about these issues.

File: src/components/GlobalNavigation/components/DesktopView/components/CloseButton/CloseButton.spec.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (hassanraza0054) [reply src/components/GlobalNavigation/components/DesktopView/components/CloseButton/CloseButton.spec.tsx:50]: Thanks a ton for that! Removed the snapshot test for now. Keen to have a discussion in the near future on visual testing potentially (which snapshot testing I believe is meant to do!).

File: src/components/GlobalNavigation/components/DesktopView/components/CollectionImage/CollectionImage.spec.tsx [hunk 9]
```diff
@@ -18,4 +18,12 @@ describe('<GlobalNavigation.DesktopView.CollectionImage />', () => {
 
     expect(results).toHaveNoViolations();
   });
+
+  it('should return null when images not provided', async () => {
```
[the remark this answers is not in the document]
Comment (hassanraza0054) [reply src/components/GlobalNavigation/components/DesktopView/components/CollectionImage/CollectionImage.spec.tsx:22]: No it does not! Great catch 👌

File: src/components/GlobalNavigation/components/DesktopView/components/CollectionItem/CollectionItem.spec.tsx [hunk 10]
```diff
@@ -1,38 +1,150 @@
...
+  it('should be accessible', async () => {
+    const { container } = render(
+      <BaseComponentWithProviders {...CollectionItemLinkFixture} />,
+    );
+    const results = await axe(container);
+
+    expect(results).toHaveNoViolations();
+  });
+
+  it('should render article meta duration', async () => {
```
[the remark this answers is not in the document]
Comment (hassanraza0054) [reply src/components/GlobalNavigation/components/DesktopView/components/CollectionItem/CollectionItem.spec.tsx:70]: image

It represents the read time for our articles! e.g. the **seven-minute read** here.

File: src/components/GlobalNavigation/components/DesktopView/components/CollectionItem/CollectionItem.spec.tsx [hunk 11]
```diff
@@ -1,38 +1,150 @@
...
+    const { container } = render(
+      <BaseComponentWithProviders {...CollectionItemArticleFixture} />,
+    );
+    const results = await axe(container);
+
+    const tag = screen.getByText(
+      (CollectionItemArticleFixture as Article).metaDuration,
```
[the remark this answers is not in the document]
Comment (hassanraza0054) [reply src/components/GlobalNavigation/components/DesktopView/components/CollectionItem/CollectionItem.spec.tsx:77]: https://testing-library.com/docs/queries/bytext/

So it takes a text value and sees if said DOM element has that matching value.

File: src/components/GlobalNavigation/components/DesktopView/components/CollectionItem/CollectionItem.spec.tsx (hunk 1 above)
Comment (hassanraza0054) [reply src/components/GlobalNavigation/components/DesktopView/components/CollectionItem/CollectionItem.spec.tsx:115]: Yeah its interesting and happy to adjust based on feedback.

Basically, tab index is either untouched (and therefore defaulted to 0) or -1 (which means not tabbable). Therefore, in these 2 situations, an untouched/tab index value of 0 is **tabbable**. That said, when tab index is 0 or not set, it doesnt exist on the dom element itself.

So test wise, the way I understood to test for that tabbability is to see if the attribute of **tabIndex** itseld didnt exist on the dom element!

File: src/components/GlobalNavigation/components/DesktopView/components/CollectionItem/CollectionItem.spec.tsx (hunk 2 above)
Comment (hassanraza0054) [reply src/components/GlobalNavigation/components/DesktopView/components/CollectionItem/CollectionItem.spec.tsx:147]: image

Unfortunately Jest fetches test id attributes with the form `data-testId` whereas our attribute is `data-test-id`.

image

We also have this same pattern in 2 other more general components.

image

What do you think would be the best way moving forward with this?

File: src/components/GlobalNavigation/components/DesktopView/components/CollectionLayout/CollectionLayout.spec.tsx (hunk 3 above)
Comment (hassanraza0054) [reply src/components/GlobalNavigation/components/DesktopView/components/CollectionLayout/CollectionLayout.spec.tsx:75]: Done!

File: src/components/GlobalNavigation/components/DesktopView/components/CloseButton/CloseButton.spec.tsx (hunk 7 above)
Comment (hassanraza0054) [reply src/components/GlobalNavigation/components/DesktopView/components/CloseButton/CloseButton.spec.tsx:17]: Good catch! You caught a legacy prop set when I wrote this test a while back. Forgot to refactor. Thanks for that

File: src/components/GlobalNavigation/components/DesktopView/components/CollectionItem/CollectionItem.spec.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (MkMan) [reply src/components/GlobalNavigation/components/DesktopView/components/CollectionItem/CollectionItem.spec.tsx:77]: I got confused by the brackets, all good 👍

File: src/components/GlobalNavigation/components/DesktopView/components/CollectionItem/CollectionItem.spec.tsx (hunk 1 above)
Comment (MkMan) [reply src/components/GlobalNavigation/components/DesktopView/components/CollectionItem/CollectionItem.spec.tsx:115]: It's a bit confusing but I see what you mean

File: src/components/GlobalNavigation/components/DesktopView/components/CollectionItem/CollectionItem.spec.tsx (hunk 2 above)
Comment (hassanraza0054) [reply src/components/GlobalNavigation/components/DesktopView/components/CollectionItem/CollectionItem.spec.tsx:147]: Ah good catch. I kept it there initially as an assertion that said link exists, but in hindsight I think I eventually meant to remove it, because that would automatically be achieved through an error being thrown by accessing an empty link. 

TLDR: agreed, should and will remove that.

**Review by MkMan — APPROVED** [review]
